neuralmonkey/decoders/decoder.py

```diff
@@ -257,26 +256,26 @@ def embedding_matrix(self) -> tf.Variable:
         if self.embeddings_source is not None:
             return self.embeddings_source.embedding_matrix
 
-        # TODO better initialization
         return tf.get_variable(
-            "word_embeddings", [len(self.vocabulary), self.embedding_size],
-            initializer=tf.random_uniform_initializer(-0.5, 0.5))
+            name="word_embeddings",
+            shape=[len(self.vocabulary), self.embedding_size],
+            initializer=tf.glorot_uniform_initializer())
 
     @tensor
     def decoding_w(self) -> tf.Variable:
         with tf.name_scope("output_projection"):
             return tf.get_variable(
-                "state_to_word_W",
-                [self.output_projection_size, len(self.vocabulary)],
-                initializer=tf.random_uniform_initializer(-0.5, 0.5))
+                name="state_to_word_W",
+                shape=[self.output_projection_size, len(self.vocabulary)],
+                initializer=tf.glorot_normal_initializer())
 
     @tensor
     def decoding_b(self) -> tf.Variable:
         with tf.name_scope("output_projection"):
             return tf.get_variable(
-                "state_to_word_b", [len(self.vocabulary)],
-                initializer=tf.constant_initializer(
-                    - math.log(len(self.vocabulary))))
+                name="state_to_word_b",
+                shape=[len(self.vocabulary)],
+                initializer=tf.zeros_initializer())
```

Comment:
Tady bych to nerušil.

Reply:
máš odkaz na něco, co to obhajuje? Experimenty s nulama běží úplně stejně jako s logem vokabuláře.

Reply:
protože mi to přijde jako šarlatánství a všichni ostatní lidi světa dávaj do všech biasů nuly kromě některých uvnitř LSTM sítí, kdy inicializace biasů jedničkou zajistí nějaký defaultní chování těch gatů.

Reply:
Ondra kdysi říkal, že to doporučuje Otrhan Firat, ale nedaří se mi to rychle vygooglit. Ale přijde mi, že tohle je jediná inicializace, která naopak není šarlatánství. Je to jediný parametr, pro který existuje jasná interpretace, co to dělá: je to apriorní pravděpodobnost slov na výstupu, a když pro ní máme jednoduchý odhad, tam mi přijde škoda toho nevyužít.

Reply:
Tady je argumentace proč tohle vyhodit:

softmax(x + b) = (e^xi * e^b) / (sum (e^xj * e^b)) = (e^b/e^b) * (e^xi / sum (e^xj)) = softmax(x)

Reply:
(https://plus.google.com/+AlbertZeyer/posts/N2wG8xuCAKZ)

Reply:
Je to __nenormalizovaná__ apriorní pravděpodobnost. Pokud by to měla bejt pravděpodobnost, necpal bys to s tim zbytkem do softmaxu, ne?

Reply:
pokud bys chtěl použít apriorní pravděpodobnosti, vynásobil bys je s výsledkem softmaxu, nevynásobil bys?

Reply:
To nemůžeš vytknou, protože to b je pro každý slovo ze slovníku jiný.

Reply:
Ta diskuze, co odkazuješ mluví o tom, jak by to vypadalo, kdybysme to inicializovali log/(1 / velikost slovníku)

Reply:
No na začátku pro všechny stejný je. Derivace budou vycházet furt stejně, protože derivace logitu vůči biasu je jedna.

Reply:
(softmax asi bude distributivní, ale jistej si nejsem - tim by neplatil muj kec začínající "pokud bys")

Reply:
Firat kterej udělal nematus tam má nuly.

Reply:
Argument o tom, že tyhle parametry jsou ty nejjednodušší na naučení mi taky přijde validní.

Reply:
ad diskuse - já vim, proto jsem jí zmínil, protože mu tam veskrze píšou, že je to jedno.

Reply:
Ale tohle je vektor, kde je v každym políčku logaritmus pravděpodobnosti toho slova, takže to nejsou stejný čísla.

Reply:
Souhlasím s tím, že zero_initializer by měl stačit. To vlákno, které Jindra postnul výše, to dobře obhajuje.

Reply:
Ještě jedna matika. Pokud začnu s biasama nastavenýma na jednu stejnou konstantu pro všechny slova _b_, pak softmax(x+b) = softmax(x). Pak dostanu update vektor _u_ takovej, že b1 = b + u. Potom softmax(x + b1) = softmax(x + b + u) = softmax(x + u), protože b je vektor stejných čísel. Víme, že gradient v biasech nezáleží na biasech, takže z toho plyne, že je to jedno.

Reply:
To je to, o čem celou dobu mluvím.

Reply:
V tom případě ti nerozumim.

Reply:
Ale to vlákno mluví o tom, když je ten vektor nastavenej na jednu stejnou konstantu!

Reply:
A jestli konstantu nebo nuly je jedno, ale jestli distribuci nebo konstantu není jedno.

Reply:
`initializer=tf.constant_initializer( BLABLABLA )` je jedna stejná konstanta

Reply:
logaritmus délky slovníku je přece pro všechny slova, sovy, králíky i princezny stejnej, ne?

Reply:
Pardón, jsem se překouk, myslel jsem, že tam jsou pravděpodobnosti.